### PR TITLE
Changes:

### PR DESCRIPTION
**Pull request by cbleslie**
- HTML Messages
- Username has color
- Root page now asks for user name before launching chat
- disabled (commented out) animals chatting. :| No frendos
= added naive templating; because sanity.

### CODE REVIEW

File: public/style.css [hunk 1]
```diff
@@ -73,13 +74,50 @@ body {
...
+.c-sign-in__form {
+  border: 1px solid var(--edge);
+  background-color: var(--inset);
+  padding: 1em;
+  display: flex;
+  flex-direction: column;
+  border-radius: 5px;
+}
+
+.c-sign-in__form > * + * {
```
Comment (cbleslie) [inline public/style.css:97]: https://alistapart.com/article/axiomatic-css-and-lobotomized-owls/

File: server.js [hunk 2]
```diff
@@ -51,37 +61,72 @@ let messageList = [];
...
+const routeGetRoot = config => (request, response) => {
+  let template = PageWrapper({
+    title: config.pageTitle,
+    children: SignIn()
+  });
+  response.send(template);
+}
+
+const routePostRoot = config => (request, response) => {
+  response.redirect(`/${request.body.user}`);
```
Comment (cbleslie) [inline server.js:83]: When you post, we make a redirect to :id with the current hidden  "user" field that's provided in the chat.

File: server.js [hunk 3]
```diff
@@ -51,37 +61,72 @@ let messageList = [];
...
-
-  subscribers.push(write);
   const cutoff = Math.max(messageList.length - 10, 0);
-
   const latest = messageList.slice(cutoff);
   if (cutoff !== 0) {
     messageList = latest;
   }
-
+  subscribers.push(write);
```
Comment (cbleslie) [inline server.js:97]: Moved this down, the above seemed like all prep.

File: server.js [hunk 4]
```diff
@@ -51,37 +61,72 @@ let messageList = [];
...
   messageList.forEach(write);
-});
+};
 
-app.post("/messages", (request, response) => {
+const routePostMessages = config => (request, response) => {
```
Comment (cbleslie) [inline server.js:101]: we curry all the routes to accept a config object. The function will return (curry) another function with the prepared configuration.

File: server.js [hunk 5]
```diff
@@ -1,21 +1,27 @@
-const express = require("express");
+const express = require('express');
+const ChatRoom = require('./templates/ChatRoom').ChatRoom;
```
Comment (cbleslie) [inline server.js:2]: The "templates" are similar to "React" in the sense that you provide props, and perhaps even other functions (templates) to the template.

**Comment by christianbundy** [comment]
WTF this is beautiful, thanks for the patch! Gonna publish this now.

File: server.js [hunk 6]
```diff
@@ -51,37 +61,72 @@ let messageList = [];
...
   messageList.forEach(write);
-});
+};
 
-app.post("/messages", (request, response) => {
+const routePostMessages = config => (request, response) => {
   if (request.body.message) {
-    publish(`me: ${request.body.message}`);
+    publish(ChatText({
+      color: stringToColor(request.body.user),
```
Comment (cbleslie) [inline server.js:104]: we get the users name and convert it into a hash to get a color.... So if you provide a unique name, you get a color, we could eventually generate SVG's in this way... or something.

File: server.js [hunk 7]
```diff
@@ -51,37 +61,72 @@ let messageList = [];
...
+  defaultUserName: 'Moose'
+}
+
+const config = {};
+// Routes
+app.get("/messages", routeGetMessages(appConfig));
+app.post("/messages", routePostMessages(appConfig));
+
+app.get("/", routeGetRoot(appConfig));
+app.post("/", routePostRoot(appConfig));
```
Comment (cbleslie) [inline server.js:125]: this is the post from the index page.
All this does is setup the chat page form/url with a user parameter that we can key off of to make a "unique" representation of a user.

A future improvement would be to get a list of subscribers by name and check tomake sure we don't have duplicates?

File: server.js [hunk 8]
```diff
@@ -51,37 +61,72 @@ let messageList = [];
...
+
+app.get("/:id", routeGetRootWithUserId(appConfig));
```
Comment (cbleslie) [inline server.js:127]: takes the ID and uses it for setting the "name" hidden text field on the chat submit form.

**Comment by cbleslie** [comment]
To test different names, just change the route name:
`http://localhost:3301/user name`
`http://localhost:3301/some other name`
etc...

**Comment by christianbundy** [comment]
come thru http://35.223.151.20:3301/

**Pull request merged**